### PR TITLE
Scopes: Work in progress

### PR DESCRIPTION
**Pull request by mdevils**

### CODE REVIEW

File: src/elements/Element.js [hunk 1]
```diff
@@ -541,9 +540,9 @@ export default class Element {
      *
      * @returns {Element}
      */
-    remove() {
+    remove(): ?Element {
         if (!this.parentElement) {
-            return this;
+            return;
```
Comment (markelog) [inline src/elements/Element.js]: Why?

File: src/elements/Element.js [hunk 2]
```diff
@@ -729,7 +726,7 @@ export default class Element {
      * @param {Element} oldElement
      */
     replaceChild(newElement: Element, oldElement: Element) {
-        return this.replaceChildren(newElement, oldElement, oldElement);
```
Comment (markelog) [inline src/elements/Element.js]: Why?

File: src/elements/Element.js [hunk 3]
```diff
@@ -842,12 +839,20 @@ export default class Element {
...
         }
+        let objectToClone = ((this: any): ConcreteElement);
+        return new objectToClone.constructor(clonedChildren);
+    }
+}
 
-        // $FlowFixMe: flow doesn't understand if this is an Element or not
-        return new this.constructor(clonedChildren);
+/**
+ * Artificial class for correct flow behaviour.
```
Comment (markelog) [inline src/elements/Element.js]: So this created only to pass `flow` checks?

File: src/elements/ElementAssert.js [hunk 4]
```diff
@@ -13,7 +13,6 @@ export default class ElementAssert {
      */
     constructor(elements: Array<Element>) {
         this._elements = elements;
-        this._currentElement = null;
```
Comment (markelog) [inline src/elements/ElementAssert.js]: This id done for the `flow` checks right?

File: src/elements/types/ArrayPattern.js [hunk 5]
```diff
@@ -15,7 +15,9 @@ export default class ArrayPattern extends Node {
                 children.moveNext();
                 children.skipNonCode();
             } else {
-                children.assertPattern();
+                if (!children.isNode('MemberExpression')) {
```
Comment (markelog) [inline src/elements/types/ArrayPattern.js]: I see you fixed some other unrelated issues here too?

Comment (markelog) [inline src/elements/types/ArrayPattern.js]: Still needs tests and not in the plugin file

File: src/plugins/scopes/ScopesApi.js [hunk 6]
```diff
@@ -0,0 +1,474 @@
+/* @flow */
+import type Element from '../../elements/Element';
```
Comment (markelog) [inline src/plugins/scopes/ScopesApi.js]: Tests?

File: src/plugins/scopes/ScopesPlugin.js [hunk 7]
```diff
@@ -0,0 +1,16 @@
+/* @flow */
+
+import BasePlugin from '../BasePlugin';
+import ScopesApi from './ScopesApi';
+import type Program from '../../elements/types/Program';
+import type Node from '../../elements/Node';
+
+export default class ScopesPlugin extends BasePlugin {
+    getPluginName(): string {
```
Comment (markelog) [inline src/plugins/scopes/ScopesPlugin.js]: Shouldn't this be a property?

File: src/utils/lines.js [hunk 8]
```diff
@@ -1,7 +1,7 @@
-/* flow */
+/* @flow */
```
Comment (markelog) [inline src/utils/lines.js]: Unrelated, but i suppose it is done in interest of time?

File: src/elements/types/MethodDefinition.js [hunk 9]
```diff
@@ -74,4 +74,8 @@ export default class MethodDefinition extends Node {
     get static() {
         return this._static;
     }
+
+    get computed() {
+        return this._computed;
```
Comment (markelog) [inline src/elements/types/MethodDefinition.js]: Tests?

File: src/elements/types/Program.js [hunk 10]
```diff
@@ -59,15 +72,42 @@ export default class Program extends Statement {
...
-    _addElementsToSearchIndex(elements: Array<Element>) {
+    _addElementsToProgram(elements: Array<Element>) {
         this._searchIndex.addElements(elements);
+        this._emit('elements-add', elements);
     }
 
-    _removeElementsFromSearchIndex(elements: Array<Element>) {
+    _removeElementsFromProgram(elements: Array<Element>) {
         this._searchIndex.removeElements(elements);
+        this._emit('elements-remove', elements);
```
Comment (markelog) [inline src/elements/types/Program.js]: Tests?

File: src/plugins/scopes/ScopesApi.js [hunk 11]
```diff
@@ -0,0 +1,474 @@
...
+            if (parentElement instanceof JSXNamespacedName) {
+                if (node === parentElement.name) {
+                    return;
+                }
+            }
+            scope._addReference({node, name, read: true, write: false});
+        }
+    }
+
+    _addIdentifier(node: Identifier) {
```
Comment (markelog) [inline src/plugins/scopes/ScopesApi.js]: I suppose we really need coverage tools implemented, otherwise it's pretty unclear

File: test/lib/elements/Element.js [hunk 12]
```diff
@@ -109,7 +109,8 @@ describe('Element', () => {
             let program = parseAndGetProgram('var first = 1; var second = 2;');
 
             expect(() => {
-                parseAndGetProgram('function foo(){ return\n; }').remove().remove();
+                program.remove();
```
Comment (markelog) [inline test/lib/elements/Element.js]: huh?

File: src/plugins/scopes/ScopesApi.js [hunk 13]
```diff
@@ -0,0 +1,474 @@
...
+        this._addElement(program);
+    }
+
+    _program: Program;
+    _scopesMap: Map<Element, Scope>;
+
+    _addElement(element: Element) {
+        if (element instanceof Node) {
+            let nodes = buildNodeList((element: Node));
+            for (let i = 0; i < nodes.length; i++) {
```
Comment (markelog) [inline src/plugins/scopes/ScopesApi.js]: `forEach` sounds cooler here

File: src/plugins/scopes/ScopesApi.js [hunk 14]
```diff
@@ -0,0 +1,474 @@
...
+    _removeElement(element: Element) {
+        if (element instanceof Node) {
+            let nodes = buildNodeList((element: Node));
+            for (let i = 0; i < nodes.length; i++) {
+                this._removeNode(nodes[i]);
+            }
+        }
+    }
+
+    _addNode(node: Node) {
```
Comment (markelog) [inline src/plugins/scopes/ScopesApi.js]: You could define a return value here, would be more checks, less errors

File: src/plugins/scopes/ScopesApi.js [hunk 15]
```diff
@@ -0,0 +1,474 @@
...
+        if (node instanceof Program) {
+            return this._addProgram(node);
+        }
+
+        if (node instanceof FunctionExpression) {
```
Comment (markelog) [inline src/plugins/scopes/ScopesApi.js]: glad you didn't go with `if..else` route here

File: src/plugins/scopes/ScopesApi.js [hunk 16]
```diff
@@ -0,0 +1,474 @@
...
+    'SwitchStatement': true,
+    'CatchClause': true
+};
+
+export default class ScopesApi {
+    constructor(program: Program) {
+        program.on('elements-add', (elements: Element[]) => {
+            elements.map(this._addElement, this);
+        });
+        program.on('elements-remove', (elements: Element[]) => {
```
Comment (markelog) [inline src/plugins/scopes/ScopesApi.js]: Since this only one statement, you could simplify it

File: src/plugins/scopes/ScopesApi.js [hunk 17]
```diff
@@ -0,0 +1,474 @@
...
+            let nodes = buildNodeList((element: Node));
+            for (let i = 0; i < nodes.length; i++) {
+                this._addNode(nodes[i]);
+            }
+        }
+    }
+
+    _removeElement(element: Element) {
+        if (element instanceof Node) {
+            let nodes = buildNodeList((element: Node));
```
Comment (markelog) [inline src/plugins/scopes/ScopesApi.js]: Everything could be `const`

File: src/plugins/scopes/ScopesApi.js [hunk 18]
```diff
@@ -0,0 +1,474 @@
...
+            node,
+            parentScope: this._getScopeFor(parentElement)
+        }));
+    }
+
+    _addJSXIdentifier(node: JSXIdentifier) {
+        let name = node.name;
+        let scope = this._getScopeFor(node);
+        let parentElement = node.parentElement;
+        if (scope && parentElement) {
```
Comment (markelog) [inline src/plugins/scopes/ScopesApi.js]: Sounds like you could simplify this whole `if`s story

File: src/plugins/BasePlugin.js [hunk 19]
```diff
@@ -0,0 +1,11 @@
+/* @flow */
+
+import type Program from '../elements/types/Program';
```
Comment (markelog) [inline src/plugins/BasePlugin.js]: Would be cool to document this stuff

File: src/plugins/scopes/Definition.js [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+    Variable: 'Variable',
+    Parameter: 'Parameter',
+    SelfReference: 'SelfReference',
+    CatchClauseError: 'CatchClauseError',
+    ImportBinding: 'ImportBinding',
+    ImplicitGlobal: 'ImplicitGlobal',
+    BuiltIn: 'BuiltIn'
+};
+
+export const typeOrder = {
```
Comment (markelog) [inline src/plugins/scopes/Definition.js]: Even with the use-case, it is pretty hard to get what and why this is

File: src/plugins/scopes/ScopesApi.js [hunk 21]
```diff
@@ -0,0 +1,474 @@
...
+    }
+
+    _addSuper(node: Super) {
+        let scope = this._getScopeFor(node);
+        if (scope) {
+            scope._addReference({node, name: 'super', read: true, write: false});
+        }
+    }
+
+    _removeNode(node: Node) {
```
Comment (markelog) [inline src/plugins/scopes/ScopesApi.js]: That is what you want to do in the next pull?

File: src/plugins/scopes/Scope.js [hunk 22]
```diff
@@ -0,0 +1,204 @@
...
+                break;
+            }
+            if (!currentScope._parentScope) {
+                let globalVariable = new Variable({
+                    name, type: types.ImplicitGlobal, scope: currentScope
+                });
+                globalVariable._addReference(reference);
+                currentScope._addVariable(globalVariable);
+                break;
+            } else {
```
Comment (markelog) [inline src/plugins/scopes/Scope.js]: Is that `else` necessary?

File: src/plugins/scopes/Scope.js [hunk 23]
```diff
@@ -0,0 +1,204 @@
...
+    _references: Map<string, Reference[]>;
+
+    _addVariable(variable: Variable) {
+        let variables = this._variables.get(variable.name);
+        if (variables) {
+            variables.push(variable);
+            variables.sort((variable1: Variable, variable2: Variable) => {
+                let typeOrder1 = typeOrder[variable1._type];
+                let typeOrder2 = typeOrder[variable2._type];
+                return typeOrder1 > typeOrder2 ? 1 : (typeOrder1 < typeOrder2 ? -1 : 0);
```
Comment (markelog) [inline src/plugins/scopes/Scope.js]: Nested ternaries? Maybe simple `if` would suffice?

File: src/elements/types/RestElement.js [hunk 24]
```diff
@@ -9,14 +9,15 @@ export default class RestElement extends Node {
     }
 
     _acceptChildren(children) {
-        let argument = null;
```
Comment (markelog) [inline src/elements/types/RestElement.js]: Tests?

File: src/elements/types/Property.js [hunk 25]
```diff
@@ -23,6 +23,7 @@ export default class Property extends Node {
             children.passToken();
             children.skipNonCode();
 
+            computed = children.isToken('Punctuator', '[');
```
Comment (markelog) [inline src/elements/types/Property.js]: Tests?

File: src/elements/types/CatchClause.js [hunk 26]
```diff
@@ -12,7 +12,7 @@ export default class CatchClause extends Node {
         children.passToken('Punctuator', '(');
         children.skipNonCode();
 
-        let param = children.passExpression();
+        let param = children.passPattern();
```
Comment (markelog) [inline src/elements/types/CatchClause.js]: Tests?

File: src/Parser.js [hunk 27]
```diff
@@ -133,7 +135,19 @@ export default class Parser {
     parse(code: string): Program {
         let ast = this._parseAst(code);
         let tokens = this._processTokens(ast, code);
-        return buildElementTree(ast, tokens);
+        let program = buildElementTree(ast, tokens);
+        let programPlugins = {};
+        let plugins = this._options.plugins;
+        for (let plugin of plugins) {
+            let api = plugin.getProgramApi(program);
```
Comment (markelog) [inline src/Parser.js]: Plugin could return falsy value for some reason? What reason could be?

File: src/Parser.js [hunk 28]
```diff
@@ -133,7 +135,19 @@ export default class Parser {
...
+            if (api) {
+                programPlugins[plugin.getPluginName()] = api;
```
Comment (markelog) [inline src/Parser.js]: So one plugin could silently re-define another?

File: src/plugins/scopes/ScopesPlugin.js [hunk 29]
```diff
@@ -0,0 +1,16 @@
+/* @flow */
+
+import BasePlugin from '../BasePlugin';
+import ScopesApi from './ScopesApi';
+import type Program from '../../elements/types/Program';
+import type Node from '../../elements/Node';
+
+export default class ScopesPlugin extends BasePlugin {
```
Comment (markelog) [inline src/plugins/scopes/ScopesPlugin.js]: Does it need to extend `BasePlugin`? Why?

File: src/plugins/scopes/ScopesPlugin.js [hunk 30]
```diff
@@ -0,0 +1,16 @@
+/* @flow */
+
+import BasePlugin from '../BasePlugin';
+import ScopesApi from './ScopesApi';
+import type Program from '../../elements/types/Program';
```
Comment (markelog) [inline src/plugins/scopes/ScopesPlugin.js]: Redundant?

File: src/plugins/scopes/ScopesPlugin.js [hunk 31]
```diff
@@ -0,0 +1,16 @@
...
+import type Node from '../../elements/Node';
```
Comment (markelog) [inline src/plugins/scopes/ScopesPlugin.js]: Redundant?

File: test/lib/plugins/scopes/ScopesPlugin.arguments.js [hunk 32]
```diff
@@ -0,0 +1,194 @@
...
+            expect(paramA.name).to.equal('a');
+            expect(paramA.type).to.equal('Parameter');
+            expect(paramA.definitions[0].type).to.equal('Parameter');
+            expect(paramA.definitions[0].node.parentElement.type).to.equal('Property');
+            expect(paramA.references[0].node.parentElement.type).to.equal('UpdateExpression');
+            expect(paramA.references[0].isReadWrite).to.equal(true);
+
+            let paramB = globalScope.childScopes[0].variables[0];
+
+            expect(paramB.name).to.equal('b');
```
Comment (markelog) [inline test/lib/plugins/scopes/ScopesPlugin.arguments.js]: `deep.equal` maybe?

File: src/Parser.js (hunk 27 above)
Comment (mdevils) [inline src/Parser.js]: Not every plugin will add Api to program (some of them will later affect parser behaviour).

File: src/elements/Element.js (hunk 1 above)
Comment (mdevils) [inline src/elements/Element.js]: In DOM `remove()` does not return anything.

File: src/elements/Element.js [hunk 33]
```diff
@@ -541,9 +540,9 @@ export default class Element {
      *
      * @returns {Element}
      */
-    remove() {
+    remove(): ?Element {
```
Comment (mdevils) [inline src/elements/Element.js]: To be removed.

File: src/elements/Element.js (hunk 2 above)
Comment (mdevils) [inline src/elements/Element.js]: Because `replaceChildren` does not return anything.

File: src/Parser.js (hunk 28 above)
Comment (mdevils) [inline src/Parser.js]: TDB: Worth throwing an error.

File: src/elements/Element.js (hunk 3 above)
Comment (mdevils) [inline src/elements/Element.js]: Yes.

File: src/elements/ElementAssert.js (hunk 4 above)
Comment (mdevils) [inline src/elements/ElementAssert.js]: Yes.

File: src/elements/types/ArrayPattern.js (hunk 5 above)
Comment (mdevils) [inline src/elements/types/ArrayPattern.js]: TBD: Test.

File: src/elements/types/CatchClause.js (hunk 26 above)
Comment (mdevils) [inline src/elements/types/CatchClause.js]: TBD: Test.

File: src/elements/types/MethodDefinition.js (hunk 9 above)
Comment (mdevils) [inline src/elements/types/MethodDefinition.js]: TBD: Test.

File: src/elements/types/Program.js (hunk 10 above)
Comment (mdevils) [inline src/elements/types/Program.js]: TBD: Test.

File: src/elements/types/Property.js (hunk 25 above)
Comment (mdevils) [inline src/elements/types/Property.js]: It is tested.

File: src/elements/types/RestElement.js (hunk 24 above)
Comment (mdevils) [inline src/elements/types/RestElement.js]: TBD: Test.

File: src/plugins/BasePlugin.js (hunk 19 above)
Comment (mdevils) [inline src/plugins/BasePlugin.js]: TBD: JsDoc.

File: src/plugins/scopes/Definition.js (hunk 20 above)
Comment (mdevils) [inline src/plugins/scopes/Definition.js]: TBD: JsDoc

File: src/plugins/scopes/Scope.js (hunk 23 above)
Comment (mdevils) [inline src/plugins/scopes/Scope.js]: TBD: Simplify.

File: src/plugins/scopes/Scope.js (hunk 22 above)
Comment (mdevils) [inline src/plugins/scopes/Scope.js]: TBD: Simplify.

File: src/plugins/scopes/ScopesApi.js (hunk 6 above)
Comment (mdevils) [inline src/plugins/scopes/ScopesApi.js]: What is missing in the tests?

File: src/plugins/scopes/ScopesApi.js (hunk 16 above)
Comment (mdevils) [inline src/plugins/scopes/ScopesApi.js]: Please elaborate on this.

File: src/plugins/scopes/ScopesApi.js (hunk 13 above)
Comment (mdevils) [inline src/plugins/scopes/ScopesApi.js]: `for` is for performance reasons here.

File: src/plugins/scopes/ScopesApi.js (hunk 17 above)
Comment (mdevils) [inline src/plugins/scopes/ScopesApi.js]: I'm not in favour of using `const` for local `currently-non-changing` variables.

File: src/plugins/scopes/ScopesApi.js (hunk 14 above)
Comment (mdevils) [inline src/plugins/scopes/ScopesApi.js]: This method does not actually return anything. Returns are just for simplicity inside.

File: src/plugins/scopes/ScopesApi.js (hunk 18 above)
Comment (mdevils) [inline src/plugins/scopes/ScopesApi.js]: TBD: Simplify.

File: src/plugins/scopes/ScopesApi.js (hunk 21 above)
Comment (mdevils) [inline src/plugins/scopes/ScopesApi.js]: Correct.

File: test/lib/elements/Element.js (hunk 12 above)
Comment (mdevils) [inline test/lib/elements/Element.js]: Please elaborate on this.

File: src/utils/lines.js (hunk 8 above)
Comment (mdevils) [inline src/utils/lines.js]: `/* flow */` is incorrect flow comment. `/* @flow */` is correct.

File: src/plugins/scopes/ScopesPlugin.js (hunk 31 above)
Comment (mdevils) [inline src/plugins/scopes/ScopesPlugin.js]: TBD: Remove.

File: src/plugins/scopes/ScopesPlugin.js (hunk 29 above)
Comment (mdevils) [inline src/plugins/scopes/ScopesPlugin.js]: For abstraction. `BasePlugin` is abstract class and every plugin defined inside `cst` package should extend it.

File: src/plugins/scopes/ScopesPlugin.js (hunk 7 above)
Comment (mdevils) [inline src/plugins/scopes/ScopesPlugin.js]: TBD: Convert to property.

File: src/utils/lines.js (hunk 8 above)
Comment (hzoo) [inline src/utils/lines.js]: just fyi: you can do `/* @noflow */` to not check a file

Comment (markelog) [inline src/utils/lines.js]: > /* flow */ is incorrect flow comment. /* @flow */ is correct.

Yeah, i was asking, why you did it here, since this is unrelated change, and i guess i answered it too :)

File: src/Parser.js (hunk 27 above)
Comment (markelog) [inline src/Parser.js]: So clearly a getter `getProgramApi` would change parser behaviour? Shouldn't it be safe, idempotent?

File: src/elements/Element.js (hunk 1 above)
Comment (markelog) [inline src/elements/Element.js]: Ohsies, that is some news, i guess DOM is still incoherent even with new methods :/

File: src/elements/Element.js (hunk 3 above)
Comment (markelog) [inline src/elements/Element.js]: :/

File: src/elements/ElementAssert.js (hunk 4 above)
Comment (markelog) [inline src/elements/ElementAssert.js]: Do we have any other flow failures? I remember i have those a lot, when i tried to fix this line

File: src/plugins/scopes/ScopesApi.js (hunk 6 above)
Comment (markelog) [inline src/plugins/scopes/ScopesApi.js]: I don't see tests for this whole file, like constructor call or anything in that nature, although it is pretty much consist with private methods/properties

File: src/plugins/scopes/ScopesApi.js (hunk 16 above)
Comment (markelog) [inline src/plugins/scopes/ScopesApi.js]: I mean you could omit parentheses. Or go even further, like 
```js
program.on('elements-remove', (elements: Element[]) => elements.map(() => this._removeElement));
```

Although i wouldn't :)

File: src/plugins/scopes/ScopesApi.js (hunk 13 above)
Comment (markelog) [inline src/plugins/scopes/ScopesApi.js]: I could say about premature optimization, but if you expect a hot path here...

File: src/plugins/scopes/ScopesApi.js (hunk 14 above)
Comment (markelog) [inline src/plugins/scopes/ScopesApi.js]: We still point that out in the method declaration and also be sure about consistent return, so no one would depend on that side-effect, if such would surface

File: src/plugins/scopes/ScopesPlugin.js (hunk 31 above)
Comment (markelog) [inline src/plugins/scopes/ScopesPlugin.js]: I wonder how many of those we have, i guess we need `disallowDeadCode` rule in JSCS

File: src/plugins/scopes/ScopesPlugin.js (hunk 29 above)
Comment (markelog) [inline src/plugins/scopes/ScopesPlugin.js]: Why? It seems as pretty basic, like i would say you should abstract something if there is something to abstract and since js doesn't have entities like interfaces or something.

Would advise you to document that somewhere in the wiki

File: test/lib/elements/Element.js (hunk 12 above)
Comment (markelog) [inline test/lib/elements/Element.js]: I just was surprised to see unrelated changes here, like in https://github.com/cst/cst/pull/88/files#r48984590

File: src/plugins/scopes/ScopesPlugin.js (hunk 31 above)
Comment (hzoo) [inline src/plugins/scopes/ScopesPlugin.js]: Unused variables or code that isn't used/called per file?

Comment (markelog) [inline src/plugins/scopes/ScopesPlugin.js]: Not sure if i can call this structure "variable", but yeah, you get the gist, naming things is hard

File: src/plugins/scopes/ScopesApi.js (hunk 6 above)
Comment (mdevils) [inline src/plugins/scopes/ScopesApi.js]: `ScopesPlugin` tests are acutally testing not `plugin`, but `ScopesApi`, will rename those tests.

File: src/plugins/scopes/ScopesPlugin.js (hunk 29 above)
Comment (mdevils) [inline src/plugins/scopes/ScopesPlugin.js]: We have two entities which support abstract classes and interfaces: `flow` and human brain.

File: src/plugins/scopes/ScopesPlugin.js (hunk 31 above)
Comment (mdevils) [inline src/plugins/scopes/ScopesPlugin.js]: Unused flow types are not tracked currently.

File: src/elements/ElementAssert.js (hunk 4 above)
Comment (mdevils) [inline src/elements/ElementAssert.js]: Nope, just this one.

File: src/plugins/scopes/Scope.js [hunk 34]
```diff
@@ -0,0 +1,204 @@
...
+    get node(): Node {
+        return this._node;
+    }
+
+    get parentScope(): ?Scope {
+        return this._parentScope;
+    }
+
+    get variables(): Variable[] {
+        return [].concat.apply([], Array.from(this._variables.values()));
```
Comment (hzoo) [inline src/plugins/scopes/Scope.js]: Oh why was this necessary rather than just `Array.from(this._variables.values()` which returns an array?

Comment (mdevils) [inline src/plugins/scopes/Scope.js]: To flatten. `[[a, b, c], [x, y, z]]` -> `[a, b, c, x, y, z]`

File: src/plugins/scopes/ScopesPlugin.js (hunk 29 above)
Comment (markelog) [inline src/plugins/scopes/ScopesPlugin.js]: I'm saying it is not a common practise, as far as i'm aware, to create empty classes with no logic to extend it afterwards, so for me, as a developer interested to write a plugin, wouldn't be a lot of sense to extend from that plugin.

At the moment at least.

File: src/plugins/scopes/ScopesApi.js (hunk 6 above)
Comment (mdevils) [inline src/plugins/scopes/ScopesApi.js]: TBD: Rename tests

File: src/plugins/scopes/Scope.js (hunk 34 above)
Comment (markelog) [inline src/plugins/scopes/Scope.js]: Maybe 

```js
Array.from(this._variables.values()).reduce((first, second) => first.concat(second))
```

Looks better?

File: src/plugins/scopes/Scope.js [hunk 35]
```diff
@@ -0,0 +1,204 @@
...
+        let variables = this._variables.get(variable.name);
+        if (variables) {
+            variables.push(variable);
+            variables.sort((variable1: Variable, variable2: Variable) => {
+                let typeOrder1 = typeOrder[variable1._type];
+                let typeOrder2 = typeOrder[variable2._type];
+                return typeOrder1 > typeOrder2 ? 1 : (typeOrder1 < typeOrder2 ? -1 : 0);
+            });
+        } else {
+            this._variables.set(variable.name, [variable]);
```
Comment (markelog) [inline src/plugins/scopes/Scope.js]: `variable.name` is a string right? So why `Map` then?

File: src/plugins/scopes/Variable.js [hunk 36]
```diff
@@ -0,0 +1,63 @@
...
+
+    _transferReferences(variable: Variable) {
+        for (let reference of this._references) {
+            this._references.delete(reference);
+            variable._references.add(reference);
+            reference._variable = variable;
+        }
+    }
+
+    get name(): string {
```
Comment (markelog) [inline src/plugins/scopes/Variable.js]: Why just not make it a simple property? Same for `type`

File: src/plugins/scopes/Scope.js (hunk 35 above)
Comment (mdevils) [inline src/plugins/scopes/Scope.js]: To be able to select all the values (`this._variables.values()`)

File: src/plugins/scopes/Variable.js (hunk 36 above)
Comment (mdevils) [inline src/plugins/scopes/Variable.js]: It should be read-only.

File: src/plugins/scopes/ScopesPlugin.js (hunk 29 above)
Comment (mdevils) [inline src/plugins/scopes/ScopesPlugin.js]: Consider this class as an interface, which is also used for `flow` support.

File: src/plugins/scopes/Scope.js (hunk 34 above)
Comment (hzoo) [inline src/plugins/scopes/Scope.js]: OH ok nvm - I guess they are all good - maybe a comment to say it's flattening if you want

Comment (mdevils) [inline src/plugins/scopes/Scope.js]: TBD: Add comment

Comment (markelog) [inline src/plugins/scopes/Scope.js]: There is a couple places like that, seems weird to add comments for both cases

File: src/plugins/scopes/ScopesPlugin.js (hunk 29 above)
Comment (markelog) [inline src/plugins/scopes/ScopesPlugin.js]: I would add a comment then.

> For plugins outside of cst it's not necessary to extend this class.

So if we move it to its own repo, it would stop extending it?

File: src/plugins/scopes/Variable.js (hunk 36 above)
Comment (markelog) [inline src/plugins/scopes/Variable.js]: You could use `defineProperty` but i suppose it isn't important

File: src/plugins/scopes/Scope.js (hunk 35 above)
Comment (markelog) [inline src/plugins/scopes/Scope.js]:
```js
Object.keys(this._variables).map(key => this._variables[key])
```
Would also work, `Map` supposed to be slower, besides `values` are cool, but `Map` have, as you probably know, a reach api, to use it only because of the `values` method
